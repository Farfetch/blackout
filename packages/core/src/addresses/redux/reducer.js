/**
 * @module addresses/reducer
 * @category Addresses
 * @subcategory Reducer
 */

import * as actionTypes from './actionTypes';
import { combineReducers } from 'redux';
import { createReducerWithResult, reducerFactory } from '../../helpers/redux';
import get from 'lodash/get';
import omit from 'lodash/omit';
import produce from 'immer';

const INITIAL_STATE = {
  error: null,
  isLoading: false,
  result: null,
  predictions: {
    result: null,
    error: null,
    isLoading: false,
  },
  predictionDetails: {
    result: null,
    error: null,
    isLoading: false,
  },
  addresses: {
    error: null,
    isLoading: false,
  },
  address: {
    error: {},
    isLoading: {},
  },
  addressSchema: {
    error: null,
    isLoading: false,
  },
  /* Used for operations related with the default address that
    have a result associated, such as getting the default contact address */
  defaultAddressDetails: {
    error: null,
    isLoading: false,
    result: null,
  },
};

export const getDefaultAddress = (addressesList, prop) => {
  for (const key in addressesList) {
    if (addressesList[key] && addressesList[key][prop]) {
      return addressesList[key];
    }
  }
  return null;
};

const result = (state = INITIAL_STATE.result, action = {}) => {
  switch (action.type) {
    case actionTypes.GET_ADDRESSES_SUCCESS:
      return action.payload.result;
    case actionTypes.CREATE_ADDRESS_SUCCESS:
      return [...state, action.meta.addressId];
    case actionTypes.DELETE_ADDRESS_SUCCESS: {
      return (
        state && state.filter(addressId => addressId !== action.meta.addressId)
      );
    }
    default:
      return state;
  }
};

const error = (state = INITIAL_STATE.error, action = {}) => {
  switch (action.type) {
    case actionTypes.GET_PREDICTION_FAILURE:
    case actionTypes.GET_PREDICTION_DETAILS_FAILURE:
    case actionTypes.GET_ADDRESS_FAILURE:
    case actionTypes.GET_ADDRESSES_FAILURE:
    case actionTypes.CREATE_ADDRESS_FAILURE:
    case actionTypes.UPDATE_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.GET_DEFAULT_CONTACT_ADDRESS_FAILURE:
      return action.payload.error;
    case actionTypes.GET_PREDICTION_REQUEST:
    case actionTypes.GET_PREDICTION_DETAILS_REQUEST:
    case actionTypes.GET_ADDRESS_REQUEST:
    case actionTypes.GET_ADDRESSES_REQUEST:
    case actionTypes.CREATE_ADDRESS_REQUEST:
    case actionTypes.UPDATE_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.GET_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_REQUEST:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const isLoading = (state = INITIAL_STATE.isLoading, action = {}) => {
  switch (action.type) {
    case actionTypes.GET_PREDICTION_REQUEST:
    case actionTypes.GET_PREDICTION_DETAILS_REQUEST:
    case actionTypes.GET_ADDRESSES_REQUEST:
    case actionTypes.GET_ADDRESS_REQUEST:
    case actionTypes.CREATE_ADDRESS_REQUEST:
    case actionTypes.UPDATE_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_REQUEST:
    case actionTypes.GET_DEFAULT_CONTACT_ADDRESS_REQUEST:
      return true;
    case actionTypes.GET_PREDICTION_FAILURE:
    case actionTypes.GET_PREDICTION_SUCCESS:
    case actionTypes.GET_PREDICTION_DETAILS_FAILURE:
    case actionTypes.GET_PREDICTION_DETAILS_SUCCESS:
    case actionTypes.GET_ADDRESSES_FAILURE:
    case actionTypes.GET_ADDRESSES_SUCCESS:
    case actionTypes.GET_ADDRESS_FAILURE:
    case actionTypes.GET_ADDRESS_SUCCESS:
    case actionTypes.CREATE_ADDRESS_FAILURE:
    case actionTypes.CREATE_ADDRESS_SUCCESS:
    case actionTypes.UPDATE_ADDRESS_FAILURE:
    case actionTypes.UPDATE_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_SUCCESS:
    case actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_SUCCESS:
    case actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_SUCCESS:
    case actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_SUCCESS:
    case actionTypes.GET_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.GET_DEFAULT_CONTACT_ADDRESS_SUCCESS:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

export const predictions = (state = INITIAL_STATE.predictions, action = {}) => {
  switch (action.type) {
    case actionTypes.GET_PREDICTION_REQUEST:
      return {
        error: INITIAL_STATE.predictions.error,
        isLoading: true,
      };
    case actionTypes.GET_PREDICTION_FAILURE:
      return {
        error: action.payload.error,
        isLoading: false,
      };
    case actionTypes.GET_PREDICTION_SUCCESS:
      return {
        error: INITIAL_STATE.predictions.error,
        isLoading: false,
        result: action.payload,
      };
    case actionTypes.RESET_PREDICTION:
      return INITIAL_STATE.predictions;
    default:
      return state;
  }
};

export const predictionDetails = createReducerWithResult(
  'GET_PREDICTION_DETAILS',
  INITIAL_STATE.predictionDetails,
  actionTypes,
);

export const entitiesMapper = {
  [actionTypes.CREATE_ADDRESS_SUCCESS]: (state, action) => {
    const id = action.payload.result;
    const createdAddress = action.payload.entities.addresses[id];

    return produce(state, draftState => {
      if (!draftState.addresses) {
        draftState.addresses = {};
      }

      draftState.addresses = {
        ...draftState.addresses,
        [id]: { ...createdAddress },
      };
    });
  },
  [actionTypes.UPDATE_ADDRESS_SUCCESS]: (state, action) => {
    const id = action.payload.result;
    const updatedAddress = action.payload.entities.addresses[id];

    return produce(state, draftState => {
      if (!draftState.addresses) {
        draftState.addresses = {};
      }

      draftState.addresses = {
        ...draftState.addresses,
        [id]: { ...updatedAddress },
      };
    });
  },
  [actionTypes.DELETE_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;
    const currentAddresses = state.addresses;

    return produce(state, draftState => {
      draftState.addresses = omit(currentAddresses, addressId);
    });
  },
  [actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;

    // Get prev default address so it can later be unmarked as the default
    const prevDefaultShippingAddress = getDefaultAddress(
      state.addresses,
      'isDefaultShippingAddress',
    );

    return produce(state, draftState => {
      if (prevDefaultShippingAddress) {
        // Unmark previous shipping address as default
        draftState.addresses[
          prevDefaultShippingAddress.id
        ].isDefaultShippingAddress = false;
      }
      // Select the selected address as default
      draftState.addresses[addressId].isDefaultShippingAddress = true;
    });
  },
  [actionTypes.SET_DEFAULT_BILLING_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;

    // Get prev default address so it can later be unmarked as the default
    const prevDefaultBillingAddress = getDefaultAddress(
      state.addresses,
      'isDefaultBillingAddress',
    );

    return produce(state, draftState => {
      if (prevDefaultBillingAddress) {
        // Unmark previous billing address as default
        draftState.addresses[
          prevDefaultBillingAddress.id
        ].isDefaultBillingAddress = false;
      }
      // Select the selected address as default
      draftState.addresses[addressId].isDefaultBillingAddress = true;
    });
  },
  [actionTypes.GET_ADDRESS_SCHEMA_SUCCESS]: (state, action) => {
    const countryId = action.payload.result;
    const countrySchema = get(
      action,
      `payload.entities.addressSchema[${countryId}]`,
      {},
    );

    return produce(state, draftState => {
      if (!draftState.addressSchema) {
        draftState.addressSchema = {};
      }

      draftState.addressSchema[countryId] = { ...countrySchema };
    });
  },
  [actionTypes.SET_DEFAULT_CONTACT_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;

    // Get prev default address so it can later be unmarked as the default
    const prevDefaultContactAddress = getDefaultAddress(
      state.addresses,
      'isPreferredAddress',
    );

    return produce(state, draftState => {
      if (prevDefaultContactAddress) {
        // Unmark previous contact address as default
        draftState.addresses[
          prevDefaultContactAddress.id
        ].isPreferredAddress = false;
      }
      // Select the selected address as default
      draftState.addresses[addressId].isPreferredAddress = true;
    });
  },
  [actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;

    return produce(state, draftState => {
      // Unmark the selected address as default
      draftState.addresses[addressId].isPreferredAddress = false;
    });
  },
  [actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;

    return produce(state, draftState => {
      if (!draftState) {
        return draftState;
      }

      const addresses = draftState.addresses;

      if (!addresses) {
        return draftState;
      }

      // Unmark the selected address as default
      const defaultBillingAddress = addresses[addressId];

      if (defaultBillingAddress) {
        defaultBillingAddress.isCurrentBilling = false;
      }

      return draftState;
    });
  },
  [actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;

    return produce(state, draftState => {
      if (!draftState) {
        return draftState;
      }

      const addresses = draftState.addresses;

      if (!addresses) {
        return draftState;
      }

      // Unmark the selected address as default
      const defaultShippingAddress = addresses[addressId];

      if (defaultShippingAddress) {
        defaultShippingAddress.isCurrentShipping = false;
      }

      return draftState;
    });
  },
};

export const address = (state = INITIAL_STATE.address, action = {}) => {
  switch (action.type) {
    case actionTypes.CREATE_ADDRESS_REQUEST:
      return {
        isLoading: { ...state.isLoading },
        error: {},
      };
    case actionTypes.GET_ADDRESS_REQUEST:
    case actionTypes.UPDATE_ADDRESS_REQUEST:
    case actionTypes.DELETE_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.addressId]: true,
        },
        error: {},
      };
    case actionTypes.GET_ADDRESS_SUCCESS:
    case actionTypes.UPDATE_ADDRESS_SUCCESS:
    case actionTypes.DELETE_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_SUCCESS:
    case actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_SUCCESS:
    case actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_SUCCESS:
    case actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.addressId]: false,
        },
      };
    case actionTypes.GET_ADDRESS_FAILURE:
    case actionTypes.UPDATE_ADDRESS_FAILURE:
    case actionTypes.DELETE_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_FAILURE:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.addressId]: false,
        },
        error: {
          ...state.error,
          [action.meta.addressId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

export const addresses = reducerFactory(
  'GET_ADDRESSES',
  INITIAL_STATE.addresses,
  actionTypes,
);

export const addressSchema = reducerFactory(
  'GET_ADDRESS_SCHEMA',
  INITIAL_STATE.addressSchema,
  actionTypes,
);

export const defaultAddressDetails = createReducerWithResult(
  'GET_DEFAULT_CONTACT_ADDRESS',
  INITIAL_STATE.defaultAddressDetails,
  actionTypes,
);

export const getError = state => state.error;
export const getIsLoading = state => state.isLoading;
export const getResult = state => state.result;

export const getPredictions = state => state.predictions;
export const getPredictionDetails = state => state.predictionDetails;
export const getAddresses = state => state.addresses;
export const getAddress = state => state.address;
export const getAddressSchema = state => state.addressSchema;
export const getDefaultAddressDetails = state => state.defaultAddressDetails;

const reducer = combineReducers({
  error,
  isLoading,
  result,
  predictions,
  predictionDetails,
  addresses,
  address,
  addressSchema,
  defaultAddressDetails,
});

/**
 * Reducer for addresses state.
 *
 * @function addressesReducer
 * @static
 *
 * @param {object} state - Current redux state.
 * @param {object} action - Action dispatched.
 *
 * @returns {object} New state.
 */
export default (state, action = {}) => {
  if (action.type === actionTypes.RESET_ADDRESSES) {
    return INITIAL_STATE;
  }

  return reducer(state, action);
};
