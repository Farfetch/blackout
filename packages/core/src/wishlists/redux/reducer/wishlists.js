import * as actionTypes from '../actionTypes';
import { combineReducers } from 'redux';
import { createMergedObject } from '../../../helpers/redux';
import wishlistsSetReducer, {
  INITIAL_STATE as SETS_INITIAL_STATE,
} from './wishlistsSets';

const INITIAL_STATE = {
  error: null,
  id: null,
  isLoading: false,
  wishlistItems: {
    error: {},
    isLoading: {},
  },
  wishlistSets: SETS_INITIAL_STATE,
};

const error = (
  state = INITIAL_STATE.error,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.ADD_ITEM_TO_WISHLIST_FAILURE:
    case actionTypes.GET_WISHLIST_FAILURE:
      return action.payload.error;
    case actionTypes.ADD_ITEM_TO_WISHLIST_REQUEST:
    case actionTypes.DELETE_WISHLIST_ITEM_REQUEST:
    case actionTypes.GET_WISHLIST_REQUEST:
    case actionTypes.UPDATE_WISHLIST_ITEM_REQUEST:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const idReducer = (
  state = INITIAL_STATE.id,
  /* istanbul ignore next */ action = {},
) => {
  if (action.type === actionTypes.GET_WISHLIST_SUCCESS) {
    return action.payload.result;
  }
  return state;
};

const isLoading = (
  state = INITIAL_STATE.isLoading,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.ADD_ITEM_TO_WISHLIST_REQUEST:
    case actionTypes.GET_WISHLIST_REQUEST:
      return true;
    case actionTypes.ADD_ITEM_TO_WISHLIST_FAILURE:
    case actionTypes.ADD_ITEM_TO_WISHLIST_SUCCESS:
    case actionTypes.GET_WISHLIST_FAILURE:
    case actionTypes.GET_WISHLIST_SUCCESS:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

const wishlistItemsReducer = (
  state = INITIAL_STATE.wishlistItems,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.DELETE_WISHLIST_ITEM_REQUEST:
    case actionTypes.UPDATE_WISHLIST_ITEM_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.wishlistItemId]: true,
        },
        error: {
          ...state.error,
          [action.meta.wishlistItemId]: null,
        },
      };
    case actionTypes.DELETE_WISHLIST_ITEM_SUCCESS:
    case actionTypes.UPDATE_WISHLIST_ITEM_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.wishlistItemId]: false,
        },
      };
    case actionTypes.DELETE_WISHLIST_ITEM_FAILURE:
    case actionTypes.UPDATE_WISHLIST_ITEM_FAILURE:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.wishlistItemId]: false,
        },
        error: {
          ...state.error,
          [action.meta.wishlistItemId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

//
// Entities mapper
//
export const entitiesMapper = {
  [actionTypes.DELETE_WISHLIST_ITEM_SUCCESS]: (
    state,
    /* istanbul ignore next */ action = {},
  ) => {
    const id = action.payload.result;
    const wishlist = action.payload.entities.wishlist[id];
    const newState = createMergedObject(state, action.payload.entities);

    newState.wishlist[id] = wishlist;

    return newState;
  },
  [actionTypes.RESET_WISHLIST_ENTITIES]: state => {
    const { wishlist, wishlistItems, wishlistSets, ...rest } = state;

    return rest;
  },
};

//
// Selectors from reducer
//
export const getError = state => state.error;
export const getId = state => state.id;
export const getIsLoading = state => state.isLoading;
export const getIsItemLoading = state => state.wishlistItems.isLoading;
export const getItemError = state => state.wishlistItems.error;

//
// Combining and exporting
//
const reducer = combineReducers({
  error,
  isLoading,
  id: idReducer,
  wishlistItems: wishlistItemsReducer,
  wishlistSets: wishlistsSetReducer,
});

/**
 * Reducer for wishlists state.
 *
 * @function wishlistsReducer
 * @static
 * @memberof module:wishlists/reducer
 *
 * @param {object} state - Current redux state.
 * @param {object} action - Action dispatched.
 *
 * @returns {object} New state.
 */
export default (state, action = {}) => {
  if (action.type === actionTypes.RESET_WISHLIST_STATE) {
    const fieldsToReset = action.payload.fieldsToReset;

    if (!fieldsToReset) {
      return INITIAL_STATE;
    } else {
      const reducerFn = (acc, field) => {
        if (state.wishlistItems[field]) {
          return {
            ...acc,
            [field]: INITIAL_STATE[field],
            wishlistItems: {
              ...acc.wishlistItems,
              [field]: INITIAL_STATE.wishlistItems[field],
            },
          };
        }

        return { ...acc, [field]: INITIAL_STATE[field] };
      };

      return fieldsToReset.reduce(reducerFn, state);
    }
  }

  return reducer(state, action);
};
