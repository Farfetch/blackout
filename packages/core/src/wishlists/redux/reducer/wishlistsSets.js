import * as actionTypes from '../actionTypes';
import { combineReducers } from 'redux';
import omit from 'lodash/omit';

export const INITIAL_STATE = {
  error: null,
  ids: null,
  isLoading: false,
  sets: {
    error: {},
    isLoading: {},
  },
};

const error = (
  state = INITIAL_STATE.error,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_WISHLIST_SETS_FAILURE:
    case actionTypes.ADD_WISHLIST_SET_FAILURE:
      return action.payload.error;
    case actionTypes.GET_WISHLIST_SETS_REQUEST:
    case actionTypes.ADD_WISHLIST_SET_REQUEST:
    case actionTypes.DELETE_WISHLIST_SET_REQUEST:
    case actionTypes.UPDATE_WISHLIST_SET_REQUEST:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const ids = (
  state = INITIAL_STATE.ids,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_WISHLIST_SETS_SUCCESS:
      return action.payload.result;
    case actionTypes.ADD_WISHLIST_SET_SUCCESS:
      return state
        ? [...state, action.payload.result]
        : [action.payload.result];
    case actionTypes.DELETE_WISHLIST_SET_SUCCESS:
      return state.filter(id => id !== action.meta.wishlistSetId);
    case actionTypes.GET_WISHLIST_SUCCESS:
      return INITIAL_STATE.ids;
    default:
      return state;
  }
};

const isLoading = (
  state = INITIAL_STATE.isLoading,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_WISHLIST_SETS_REQUEST:
    case actionTypes.ADD_WISHLIST_SET_REQUEST:
      return true;
    case actionTypes.GET_WISHLIST_SETS_SUCCESS:
    case actionTypes.GET_WISHLIST_SETS_FAILURE:
    case actionTypes.ADD_WISHLIST_SET_SUCCESS:
    case actionTypes.ADD_WISHLIST_SET_FAILURE:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

const sets = (
  state = INITIAL_STATE.sets,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.DELETE_WISHLIST_SET_REQUEST:
    case actionTypes.GET_WISHLIST_SET_REQUEST:
    case actionTypes.UPDATE_WISHLIST_SET_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.wishlistSetId]: true,
        },
        error: {
          ...state.error,
          [action.meta.wishlistSetId]: null,
        },
      };
    case actionTypes.DELETE_WISHLIST_SET_SUCCESS:
    case actionTypes.GET_WISHLIST_SET_SUCCESS:
    case actionTypes.UPDATE_WISHLIST_SET_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.wishlistSetId]: false,
        },
      };
    case actionTypes.DELETE_WISHLIST_SET_FAILURE:
    case actionTypes.GET_WISHLIST_SET_FAILURE:
    case actionTypes.UPDATE_WISHLIST_SET_FAILURE:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.wishlistSetId]: false,
        },
        error: {
          ...state.error,
          [action.meta.wishlistSetId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

//
// Entities mapper
//
export const entitiesMapper = {
  [actionTypes.DELETE_WISHLIST_SET_SUCCESS]: (
    state,
    { meta: { wishlistSetId } },
  ) => {
    const { wishlistSets, ...rest } = state;

    return {
      ...rest,
      wishlistSets: omit(wishlistSets, wishlistSetId),
    };
  },
};

//
// Selectors from reducer
//
export const getError = state => state.error;
export const getIds = state => state.ids;
export const getIsLoading = state => state.isLoading;
export const getSetError = state => state.sets.error;
export const getIsSetLoading = state => state.sets.isLoading;

//
// Combining and exporting
//
export const reducer = combineReducers({
  error,
  ids,
  isLoading,
  sets,
});

/**
 * Reducer for wishlistsSets state.
 *
 * @function wishlistsSetsReducer
 * @static
 * @memberof module:wishlists/reducer
 *
 * @param {object} state - Current redux state.
 * @param {object} action - Action dispatched.
 *
 * @returns {object} New state.
 */
export default (state, action = {}) => {
  if (action.type === actionTypes.RESET_WISHLIST_SETS_STATE) {
    const fieldsToReset = action.payload.fieldsToReset;

    if (!fieldsToReset) {
      return INITIAL_STATE;
    } else {
      const reducerFn = (acc, field) => {
        if (state.sets[field]) {
          return {
            ...acc,
            [field]: INITIAL_STATE[field],
            sets: {
              ...acc.sets,
              [field]: INITIAL_STATE.sets[field],
            },
          };
        }

        return { ...acc, [field]: INITIAL_STATE[field] };
      };

      return fieldsToReset.reduce(reducerFn, state);
    }
  }

  return reducer(state, action);
};
