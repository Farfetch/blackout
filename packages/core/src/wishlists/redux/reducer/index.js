/**
 * @module wishlists/reducer
 * @category Wishlists
 * @subcategory Reducer
 */

import { entitiesMapper as wishlistsSetsEntitiesMapper } from './wishlistsSets';
import wishlistsReducer, {
  entitiesMapper as wishlistsEntitiesMapper,
} from './wishlists';

export const entitiesMapper = {
  ...wishlistsEntitiesMapper,
  ...wishlistsSetsEntitiesMapper,
};

export default wishlistsReducer;
