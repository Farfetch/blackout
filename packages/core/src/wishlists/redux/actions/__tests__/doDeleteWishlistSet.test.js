import { doDeleteWishlistSet } from '../';
import { mockStore } from '../../../../../tests';
import {
  mockWishlistId,
  mockWishlistSetId,
  mockWishlistsResponse,
} from 'tests/__fixtures__/wishlists';
import find from 'lodash/find';
import reducer, { actionTypes } from '../../';
import thunk from 'redux-thunk';

const mockMiddlewares = [
  thunk.withExtraArgument({
    getOptions: () => ({ productImgQueryParam: '?c=2' }),
  }),
];

const wishlistMockStore = (state = {}) =>
  mockStore({ wishlist: reducer() }, state, mockMiddlewares);

describe('doDeleteWishlistSet() action creator', () => {
  let store;
  const deleteWishlistsSet = jest.fn();
  const action = doDeleteWishlistSet(deleteWishlistsSet);
  const expectedConfig = undefined;

  beforeEach(() => {
    jest.clearAllMocks();

    store = wishlistMockStore({
      wishlist: { id: mockWishlistId },
    });
  });

  it('should create the correct actions for when the delete wishlist set procedure fails', async () => {
    const expectedError = new Error('delete wishlist set error');

    deleteWishlistsSet.mockRejectedValueOnce(expectedError);

    expect.assertions(4);

    try {
      await store.dispatch(action(mockWishlistSetId));
    } catch (error) {
      expect(error).toBe(expectedError);
      expect(deleteWishlistsSet).toHaveBeenCalledTimes(1);
      expect(deleteWishlistsSet).toHaveBeenCalledWith(
        mockWishlistId,
        mockWishlistSetId,
        expectedConfig,
      );
      expect(store.getActions()).toEqual(
        expect.arrayContaining([
          {
            meta: { wishlistSetId: mockWishlistSetId },
            type: actionTypes.DELETE_WISHLIST_SET_REQUEST,
          },
          {
            meta: { wishlistSetId: mockWishlistSetId },
            payload: { error: expectedError },
            type: actionTypes.DELETE_WISHLIST_SET_FAILURE,
          },
        ]),
      );
    }
  });

  it('should create the correct actions for when the delete wishlist set procedure is successful', async () => {
    deleteWishlistsSet.mockResolvedValueOnce(mockWishlistsResponse);
    await store.dispatch(action(mockWishlistSetId));

    const actionResults = store.getActions();

    expect(deleteWishlistsSet).toHaveBeenCalledTimes(1);
    expect(deleteWishlistsSet).toHaveBeenCalledWith(
      mockWishlistId,
      mockWishlistSetId,
      expectedConfig,
    );
    expect(actionResults).toMatchObject([
      {
        meta: { wishlistSetId: mockWishlistSetId },
        type: actionTypes.DELETE_WISHLIST_SET_REQUEST,
      },
      {
        meta: { wishlistSetId: mockWishlistSetId },
        type: actionTypes.DELETE_WISHLIST_SET_SUCCESS,
      },
    ]);
    expect(
      find(actionResults, {
        type: actionTypes.DELETE_WISHLIST_SET_SUCCESS,
      }),
    ).toMatchSnapshot('delete wishlist set success payload');
  });
});
