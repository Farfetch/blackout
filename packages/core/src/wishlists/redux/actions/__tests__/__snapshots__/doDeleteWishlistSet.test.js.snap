// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doDeleteWishlistSet() action creator should create the correct actions for when the delete wishlist set procedure is successful: delete wishlist set success payload 1`] = `
Object {
  "meta": Object {
    "wishlistSetId": "77fae745-7608-4b7d-8e7d-4f6923e032ef",
  },
  "type": "@farfetch/blackout-core/DELETE_WISHLIST_SET_SUCCESS",
}
`;
